safeweb/http.go

```diff
@@ -74,25 +74,53 @@ import (
 	crand "crypto/rand"
 	"fmt"
 	"log"
+	"maps"
 	"net"
 	"net/http"
 	"net/url"
 	"path"
+	"slices"
 	"strings"
 
 	"github.com/gorilla/csrf"
 )
 
-// The default Content-Security-Policy header.
-var defaultCSP = strings.Join([]string{
-	`default-src 'self'`,      // origin is the only valid source for all content types
-	`script-src 'self'`,       // disallow inline javascript
-	`frame-ancestors 'none'`,  // disallow framing of the page
-	`form-action 'self'`,      // disallow form submissions to other origins
-	`base-uri 'self'`,         // disallow base URIs from other origins
-	`block-all-mixed-content`, // disallow mixed content when serving over HTTPS
-	`object-src 'self'`,       // disallow embedding of resources from other origins
-}, "; ")
+// CSP is the value of a Content-Security-Policy header. Keys are CSP
+// directives (like "default-src") and values are source expressions (like
+// "'self'" or "https://tailscale.com"). A nil slice value is allowed for some
+// directives like "upgrade-insecure-requests" that don't expect a list of
+// source definitions.
+type CSP map[string][]string
```

Comment:
Some general API thoughts, please feel free to disregard these.

- It would be nice if this could have some helper methods like `http.Header` does, for simple common cases like "there is only one value" (e.g., `csp.Set("default-src", "blah")`, `csp.Add("form-action", "blah')`). That way the caller doesn't always have to think about it being a sequence.
- Is there some way we can avoid the caller having to remember to put in the single quotes for values like `'self'` and `'none'`? Maybe some predefined constants, or perhaps helpers for those very specific common cases?

Reply:
I like it! Added `Add`, `Set` and `Del`.

For quoting - added some implicit logic during encoding. There's a long list of possible special values that need quoting, but importantly none of them are URLs. So I added a lazy check of "does this value contain a dot (is a URL)?". WDYT?

Reply:
That seems reasonable to me.